src/System/TimedScopes/CorrelationData.cs

```diff
@@ -0,0 +1,547 @@
+			{
+				return null;
+			}
+			return (string)m_additionalData[key];
+		}
+
+
+		/// <summary>
+		/// Add correlation data
+		/// </summary>
+		/// <param name="key">key to get data for</param>
+		/// <param name="value">value of data</param>
+		/// <remarks>Calling this method only modifies the in memory representation
+		/// of the correlation data, to modify the actual correlation data, use
+		/// Correlation.CorrelationAdd(string key, string value).</remarks>
+		/// <exception cref="ArgumentNullException"><paramref name="key"/> or <paramref name="value"/> is null</exception>
+		public void AddData(string key, string value)
+		{
+			if (key == null)
+			{
+				throw new ArgumentNullException("key");
+			}
+			if (value == null)
+			{
+				throw new ArgumentNullException("value");
+			}
+			if (string.Equals(key, TransactionIdKey, StringComparison.OrdinalIgnoreCase))
+			{
+				// Caching the TransactionId specifically as it is likely to be queried for
+				// repeatedly
+				uint transactionId;
+				if (uint.TryParse(value, out transactionId))
+				{
+					TransactionId = transactionId;
+				}
+				else
+				{
+					TransactionId = 0;
+				}
+			}
+			if (m_additionalData == null)
+			{
+				m_additionalData = new ListDictionary(StringComparer.OrdinalIgnoreCase);
+			}
+
+			m_additionalData[key] = value;
+		}
+
+
+		/// <summary>
+		/// Constructor.
+		/// </summary>
+		/// <param name="cachedUlsEventsReplayed">Have previously cached ULS events for this correlation already been replayed?</param>
+		public CorrelationData(bool cachedUlsEventsReplayed = false)
+		{
+			m_cachedUlsEventsReplayed = cachedUlsEventsReplayed ? 1 : 0;
+		}
+
+
+		/// <summary>
+		/// Constructor.
+		/// </summary>
+		/// <param name="logEventCache">Log event cache</param>
+		/// <param name="cachedUlsEventsReplayed">Have previously cached ULS events for this correlation already been replayed?</param>
+		public CorrelationData(ILogEventCache logEventCache, bool cachedUlsEventsReplayed = false)
+		{
+			Code.ExpectsArgument(logEventCache, nameof(logEventCache), TaggingUtilities.ReserveTag(0));
+			m_logEventCache = logEventCache;
+			m_cachedUlsEventsReplayed = cachedUlsEventsReplayed ? 1 : 0;
+		}
+	}
+
+
+	/// <summary>
+	/// Correlation data extension methods
+	/// </summary>
+	public static class CorrelationDataExtensionMethods
+	{
+		/// <summary>
+		/// Clone the correlation
+		/// </summary>
+		/// <param name="existingCorrelation">existing correlation</param>
+		/// <returns>cloned correlation, null if existing correlation is null</returns>
+		/// <remarks>Added as a extension method instead of a method on the
+		/// object to allow for cloning of the data when it is null, i.e.
+		/// CorrelationData data = null; data.Clone(); will not throw exception.</remarks>
+		public static CorrelationData Clone(this CorrelationData existingCorrelation)
+		{
+			if (existingCorrelation == null)
+			{
+				return null;
+			}
+
+			// TODO(lukak): VSO 323303
```

Comment:
>// TODO(lukak): VSO 323303 [](start = 3, length = 26)

please remove and substitute with a GitHub issue if PBI 323303 is still available and valid